python/cugraph/cugraph/gnn/data_loading/dist_sampler.py

```diff
@@ -0,0 +1,349 @@
+
+from cugraph.gnn.data_loading.bulk_sampler_io import create_df_from_disjoint_arrays
+
+# PyTorch is NOT optional but this is required for container builds.
+torch = import_optional("torch")
+
+TensorType = Union[torch.Tensor, cupy.ndarray, cudf.Series]
+
+
+class DistSampleWriter:
+    def __init__(
+        self,
+        directory: str,
+        *,
+        batches_per_partition: int = 256,
+        format: str = "parquet",
+    ):
+        if format != "parquet":
+            raise ValueError("Invalid format (currently supported: 'parquet')")
+
+        self.__format = format
+        self.__directory = directory
+        self.__batches_per_partition = batches_per_partition
+
+    @property
+    def _format(self):
+        return self.__format
+
+    @property
+    def _directory(self):
+        return self.__directory
+
+    @property
+    def _batches_per_partition(self):
+        return self.__batches_per_partition
+
+    def __write_minibatches_coo(self, minibatch_dict):
+        has_edge_ids = minibatch_dict["edge_id"] is not None
+        has_edge_types = minibatch_dict["edge_type"] is not None
+        has_weights = minibatch_dict["weight"] is not None
+
+        if minibatch_dict["renumber_map"] is None:
+            raise ValueError(
+                "Distributed sampling without renumbering is not supported"
+            )
+
+        fanout_length = (len(minibatch_dict["label_hop_offsets"]) - 1) // len(
+            minibatch_dict["batch_id"]
+        )
+        rank_batch_offset = minibatch_dict["batch_id"][0]
+
+        for p in range(
+            0, int(ceil(len(minibatch_dict["batch_id"]) / self.__batches_per_partition))
+        ):
+            partition_start = p * (self.__batches_per_partition)
+            partition_end = (p + 1) * (self.__batches_per_partition)
+
+            label_hop_offsets_array_p = minibatch_dict["label_hop_offsets"][
+                partition_start * fanout_length : partition_end * fanout_length + 1
+            ]
+
+            batch_id_array_p = minibatch_dict["batch_id"][partition_start:partition_end]
+            start_batch_id = batch_id_array_p[0] - rank_batch_offset
+
+            start_ix, end_ix = label_hop_offsets_array_p[[0, -1]]
+            majors_array_p = minibatch_dict["majors"][start_ix:end_ix]
+            minors_array_p = minibatch_dict["minors"][start_ix:end_ix]
+            edge_id_array_p = (
+                minibatch_dict["edge_id"][start_ix:end_ix]
+                if has_edge_ids
+                else cupy.array([], dtype="int64")
+            )
+            edge_type_array_p = (
+                minibatch_dict["edge_type"][start_ix:end_ix]
+                if has_edge_types
+                else cupy.array([], dtype="int32")
+            )
+            weight_array_p = (
+                minibatch_dict["weight"][start_ix:end_ix]
+                if has_weights
+                else cupy.array([], dtype="float32")
+            )
```

Comment:
So, are we assuming that edge IDs are always int64, edge types are always int32 (this may make sense), and edge weights are always float32?

Reply:
I think C++ only supports int32 edge types.  For edge id, the frameworks always use int64.  In any case, this does actually allow for int32 edge ids, we just have to set a default empty dtype, and I just went with int64.

Reply:
Same for edge weights, I'm not technically restricting them to float32 here.